src/main/java/net/minecraftforge/fml/common/ModTrackingList.java

```diff
@@ -0,0 +1,316 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.fml.common;
+
+import com.google.common.collect.ForwardingIterator;
+import com.google.common.collect.ForwardingList;
+import com.google.common.collect.ForwardingListIterator;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.RandomAccess;
+
+/**
+ * Wraps a list and keeps track of which active mod container added each element in the list.
+ * Adds the method {@link #getModContainer(Object)}.
+ */
+public class ModTrackingList<T> extends ForwardingList<T> implements RandomAccess
+{
+    private final List<T> delegate;
+    private final Map<T, ModContainer> modContainerMap;
+
+    public <D extends List<T> & RandomAccess> ModTrackingList(@Nonnull D delegate)
+    {
+        this(delegate, new IdentityHashMap<T, ModContainer>());
+    }
+
+    private ModTrackingList(@Nonnull List<T> delegate, @Nonnull Map<T, ModContainer> modContainerMap)
+    {
+        this.delegate = delegate;
+        this.modContainerMap = modContainerMap;
+    }
+
+    @Nullable
+    public ModContainer getModContainer(@Nonnull T element)
+    {
+        return modContainerMap.get(element);
+    }
+
+    @Override
+    protected List<T> delegate()
+    {
+        return delegate;
+    }
+
+    private void trackModContainer(@Nonnull T element)
+    {
+        ModContainer modContainer = Loader.instance().activeModContainer();
+        if (modContainer != null)
+        {
+            modContainerMap.put(element, modContainer);
+        }
+    }
+
+    @Override
+    public boolean add(@Nonnull T t)
+    {
+        boolean changed = delegate.add(t);
+        if (changed)
+        {
+            trackModContainer(t);
+        }
+        return changed;
+    }
+
+    @Override
+    public void add(int index, @Nonnull T element)
+    {
+        delegate.add(index, element);
+        trackModContainer(element);
+    }
+
+    @Override
+    public boolean addAll(int index, @Nonnull Collection<? extends T> elements)
+    {
+        boolean changed = delegate.addAll(index, elements);
+        if (changed)
+        {
+            ModContainer modContainer = Loader.instance().activeModContainer();
```

Comment:
Use something like `trackModContainer(@Nonnull Collection<? extends T> elements)` for both `addAll` methods, like you did for `add`?